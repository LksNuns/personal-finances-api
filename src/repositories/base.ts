import {
  AbstractRepository,
  FindManyOptions,
  QueryRunner,
  SelectQueryBuilder,
} from 'typeorm';

export class Base<Entity> extends AbstractRepository<Entity> {
  // TODO See better way to expose some methods from this.repository (delegator)
  createQueryBuilder(
    alias?: string,
    queryRunner?: QueryRunner
  ): SelectQueryBuilder<Entity> {
    return this.repository.createQueryBuilder(alias, queryRunner);
  }

  async findAll(): Promise<Entity[]> {
    return this.repository.find();
  }

  async find(options?: FindManyOptions<Entity>): Promise<Entity[]> {
    return this.repository.find(options);
  }

  async findById(id: string): Promise<Entity | undefined> {
    return await this.repository.findOne(id);
  }

  async destroy(id: string): Promise<Entity | undefined> {
    const resource = await this.repository.findOne(id);

    await this.repository.delete(id);

    return resource;
  }

  async count(options?: FindManyOptions<Entity>): Promise<number> {
    return this.repository.count(options);
  }
}
