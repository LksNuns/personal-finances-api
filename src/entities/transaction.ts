import { IsNotEmpty, IsPositive } from 'class-validator';

import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
} from 'typeorm';

export enum TypeEnum {
  income = 'income',
  outcome = 'outcome',
}

@Entity('transactions')
export class Transaction {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  description: string;

  @IsPositive()
  @Column('decimal', { precision: 5, scale: 2 })
  value: number;

  @IsNotEmpty()
  @Column('enum', { enum: TypeEnum })
  type: TypeEnum;

  @IsNotEmpty()
  @Column({ type: 'date' })
  executed_at: Date;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;
}
