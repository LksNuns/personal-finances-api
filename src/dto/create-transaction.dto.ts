import { TypeEnum } from '@/entities/transaction';
import { IsDateString, IsEnum, IsNotEmpty, IsPositive } from 'class-validator';

export class CreateTransactionDto {
  description: string;

  @IsPositive()
  value: number;

  @IsNotEmpty()
  @IsEnum(TypeEnum)
  type: TypeEnum;

  @IsNotEmpty()
  @IsDateString()
  executedAt: string;
}
