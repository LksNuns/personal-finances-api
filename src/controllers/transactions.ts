import { CreateTransactionDto } from '@/dto/create-transaction.dto';
import { TransactionRepository } from '@/repositories/transactions';
import { Body, Post, JsonController } from 'routing-controllers';
import { getCustomRepository } from 'typeorm';

@JsonController('/transactions')
export class TransactionsController {
  @Post('')
  public async create(@Body() data: CreateTransactionDto): Promise<any> {
    const transactionRepository = getCustomRepository(TransactionRepository);

    try {
      await transactionRepository.create(data);
    } catch (error) {}

    return { ok: 'ok' };
  }
}
