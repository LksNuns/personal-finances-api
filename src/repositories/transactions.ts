import { CreateTransactionDto } from '@/dto/create-transaction.dto';
import { Transaction } from '@/entities/transaction';
import { validate } from 'class-validator';
import { EntityRepository } from 'typeorm';
import { Base } from './base';

@EntityRepository(Transaction)
export class TransactionRepository extends Base<Transaction> {
  async create(data: CreateTransactionDto): Promise<Transaction> {
    const transaction = this.repository.create(data);

    const errors = await validate(transaction);

    if (errors.length > 0) {
      return Promise.reject(errors);
    }

    await this.repository.save(transaction);

    return Promise.resolve(transaction);
  }
}
