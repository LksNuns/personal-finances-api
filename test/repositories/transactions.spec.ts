import { CreateTransactionDto } from '@/dto/create-transaction.dto';
import { TypeEnum } from '@/entities/transaction';
import { TransactionRepository } from '@/repositories/transactions';
import { getCustomRepository } from 'typeorm';

describe('TransactionRepository', () => {
  let repository: TransactionRepository;

  beforeAll(() => {
    repository = getCustomRepository(TransactionRepository);
  });

  describe('#create', () => {
    // Partial to permit invalid params
    let params: Partial<CreateTransactionDto>;

    describe('with invalid params', () => {
      beforeEach(() => {
        // Missing required params: "type" and "executedAt"
        params = { value: 22.0 };
      });

      it('returns an error', async () => {
        await expect(
          repository.create(params as CreateTransactionDto)
        ).rejects.toMatchObject([
          { property: 'type' },
          { property: 'executedAt' },
        ]);
      });
    });

    describe('with valid params', () => {
      beforeEach(() => {
        params = {
          value: 22.0,
          type: TypeEnum.income,
          executedAt: '2020-10-10',
        };
      });

      it('saves to database', async () => {
        // Sanity check
        expect(await repository.count()).toBe(0);

        // TODO Should we mock `respository.save()` to avoid using database (?)
        const transaction = await repository.create(
          params as CreateTransactionDto
        );

        expect(await repository.count()).toBe(1);
        expect(transaction).toMatchObject({ value: 22.0 });
      });
    });
  });
});
