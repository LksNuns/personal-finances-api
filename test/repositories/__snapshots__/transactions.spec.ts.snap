// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionRepository #create with invalid params returns an error 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isNotEmpty": "type should not be empty",
    },
    "property": "type",
    "target": Transaction {
      "value": 22,
    },
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isNotEmpty": "executedAt should not be empty",
    },
    "property": "executedAt",
    "target": Transaction {
      "value": 22,
    },
    "value": undefined,
  },
]
`;

exports[`TransactionRepository #create with valid params returns an error 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isNotEmpty": "type should not be empty",
    },
    "property": "type",
    "target": Transaction {
      "value": 22,
    },
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isNotEmpty": "executedAt should not be empty",
    },
    "property": "executedAt",
    "target": Transaction {
      "value": 22,
    },
    "value": undefined,
  },
]
`;
